### PR TITLE
React auth implementation using new JWT flow

### PR DESCRIPTION
**Pull request by humphd**
This implements the basics of what's needed in the front-end in order to use the new auth flow from #1796.  It's not possible to test this without also having the back-end auth service, so I might wait until it lands to get reviews, unless you want to merge that branch into this one locally.

This works by handling redirection on login/logout to the auth service and extracting the `?access_token=...` from the URL if it exists.

### CODE REVIEW

File: src/web/src/components/Login/LoggedOut.tsx [hunk 1]
```diff
@@ -1,20 +1,11 @@
...
-      button={{
-        href: loginUrl,
-        title: 'Log In',
-        ariaLabel: 'log in',
-        Icon: ExitToAppIcon,
-      }}
-    />
-  );
+  const { login } = useAuth();
+  return <HeaderButton title="Log In" onClick={() => login()} Icon={ExitToAppIcon} />;
```
Comment (birtony) [inline src/web/src/components/Login/LoggedOut.tsx:8]: Can't we just do `onClick={login}`?

**Review by birtony — COMMENTED** [review]
I really like the auth architecture, good job @humphd! Should the `components/MyFeeds.tsx` also be updated to use the new authentication hook? Otherwise, this PR would crash our staging environment, becasue Vercel preview is throwing this error:
image

File: src/web/src/components/Login/LoggedOut.tsx (hunk 1 above)
Comment (humphd) [reply src/web/src/components/Login/LoggedOut.tsx:8]: Huh, this is weird that I didn't get this locally.  Thanks a lot for catching it, @birtony: review works!  I'll make a fix and push again.

**Review by birtony — APPROVED** [review]

**Review by manekenpix — DISMISSED** [review]

**Comment by humphd** [comment]
Rebased this and had to change quite a bit.  Can I get another review?

File: src/web/src/components/AuthProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+export interface AuthContextInterface {
+  login: (returnTo?: string) => void;
+  logout: (returnTo?: string) => void;
+  user?: User;
+  token?: string;
+}
+
+const AuthContext = createContext<AuthContextInterface>({
+  login() {
+    throw new Error('You need to wrap your component in <AuthProvider>');
```
Comment (birtony) [inline src/web/src/components/AuthProvider.tsx:19]: We don't want to do login or logout on the server-side so we are throwing here?

**Comment by humphd** [comment]
@birtony that code is sort of confusing.  With React Contexts, you have to provide a default object to use until you have your runtime values.  You'd never call this function directly.  Instead, you'd use the one we provide when we setup the AuthProvider component and use our own runtime implementation.

This is the equivalent of having a method in a base class that needs to be overridden by a sub-class that will provide its own implementation strategy.

**Comment by humphd** [comment]
Can I get some reviews on this soon?  I'd like to land it now that the back-end is in.

**Review by birtony — DISMISSED** [review]

**Comment by humphd** [comment]
@DukeManh I'm not sure how to make eslint happy with that style.  If you know, I'll change it.

File: src/web/src/User.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  // TODO: more user details...
+  id: string;
+
+  email: string;
+
+  name: string;
+
+  isAdmin: boolean;
+
+  constructor(id: string, email: string, name: string, isAdmin: boolean) {
```
[the remark this answers is not in the document]
Comment (DukeManh) [reply src/web/src/User.ts:13]: > @DukeManh I'm not sure how to make eslint happy with that style. If you know, I'll change it.
I think these will do it. 
`
    /**
     *  Allow typescript empty constructor
     *  https://eslint.org/docs/rules/no-empty-function#options
     *  https://eslint.org/docs/rules/no-useless-constructor
     */
    'no-useless-constructor': 'off',
    'no-empty-function': ['error', { allow: ['constructors'] }],
`

**Comment by humphd** [comment]
@DukeManh done, and rebased.  Can I get another review?

**Review by DukeManh — APPROVED** [review]

**Review by Metropass — APPROVED** [review]

**Pull request merged**